File: scripts/verify_rational_grid_vectors.mjs
#!/usr/bin/env node

import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const REPOSITORY_ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const DEFAULT_VECTOR_PATH = resolve(
  REPOSITORY_ROOT,
  "conformance",
  "rational-grid-canonicalization-v1.json",
);
const CANONICAL_DECIMAL = /^(?:0|-[1-9][0-9]*|[1-9][0-9]*)$/;
const LOWER_HEX_OCTETS = /^(?:[0-9a-f]{2})+$/;
const INT64_MIN = -(2n ** 63n);
const INT64_MAX = 2n ** 63n - 1n;

class VectorVerificationError extends Error {}

function requireObject(value, label) {
  if (value === null || Array.isArray(value) || typeof value !== "object") {
    throw new VectorVerificationError(`${label} must be a JSON object`);
  }
  return value;
}

function requireArray(value, label) {
  if (!Array.isArray(value)) {
    throw new VectorVerificationError(`${label} must be a JSON array`);
  }
  return value;
}

function requireText(value, label) {
  if (typeof value !== "string" || value.length === 0) {
    throw new VectorVerificationError(`${label} must be a nonempty string`);
  }
  return value;
}

function integer(value, label) {
  const text = requireText(value, label);
  if (!CANONICAL_DECIMAL.test(text)) {
    throw new VectorVerificationError(`${label} must use canonical base-10 integer text`);
  }
  return BigInt(text);
}

function int64(value, label) {
  const parsed = integer(value, label);
  if (parsed < INT64_MIN || parsed > INT64_MAX) {
    throw new VectorVerificationError(`${label} must fit signed int64`);
  }
  return parsed;
}

function assertUnicodeScalarString(value) {
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) {
        throw new VectorVerificationError("RFC8785 input contains an unpaired surrogate");
      }
      index += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      throw new VectorVerificationError("RFC8785 input contains an unpaired surrogate");
    }
  }
}

function canonicalJson(value) {
  if (value === null || typeof value === "boolean") {
    return JSON.stringify(value);
  }
  if (typeof value === "string") {
    assertUnicodeScalarString(value);
    return JSON.stringify(value);
  }
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new VectorVerificationError("RFC8785 input contains a non-finite number");
    }
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => canonicalJson(item)).join(",")}]`;
  }
  if (typeof value === "object") {
    const fields = Object.keys(value).sort();
    return `{${fields
      .map((key) => {
        assertUnicodeScalarString(key);
        return `${JSON.stringify(key)}:${canonicalJson(value[key])}`;
      })
      .join(",")}}`;
  }
  throw new VectorVerificationError(`value of type ${typeof value} is not JSON`);
}

function sha256(bytes) {
  return createHash("sha256").update(bytes).digest("hex");
}

function absolute(value) {
  return value < 0n ? -value : value;
}

function gcd(left, right) {
  let a = absolute(left);
  let b = absolute(right);
  while (b !== 0n) {
    const remainder = a % b;
    a = b;
    b = remainder;
  }
  return a;
}

function floorDivmod(numerator, denominator) {
  let quotient = numerator / denominator;
  let remainder = numerator % denominator;
  if (remainder < 0n) {
    quotient -= 1n;
    remainder += denominator;
  }
  return [quotient, remainder];
}

function roundHalfEven(numerator, denominator) {
  if (denominator <= 0n) {
    throw new VectorVerificationError("rounding denominator must be positive");
  }
  const [quotient, remainder] = floorDivmod(numerator, denominator);
  const doubledRemainder = remainder * 2n;
  if (doubledRemainder < denominator) return quotient;
  if (doubledRemainder > denominator) return quotient + 1n;
  return quotient % 2n === 0n ? quotient : quotient + 1n;
}

function createGrid(input, label) {
  const fields = requireObject(input.grid, `${label}.grid`);
  const origin = int64(fields.grid_origin_ns, `${label}.grid.grid_origin_ns`);
  let rateNum = integer(fields.rate_num, `${label}.grid.rate_num`);
  let rateDen = integer(fields.rate_den, `${label}.grid.rate_den`);
  if (rateNum <= 0n || rateDen <= 0n) {
    throw new VectorVerificationError(`${label}.grid rate must be positive`);
  }
  const rateDivisor = gcd(rateNum, rateDen);
  rateNum /= rateDivisor;
  rateDen /= rateDivisor;
  const rawPeriodNumerator = 1_000_000_000n * rateDen;
  const periodDivisor = gcd(rawPeriodNumerator, rateNum);
  return {
    origin,
    periodNum: rawPeriodNumerator / periodDivisor,
    periodDen: rateNum / periodDivisor,
  };
}

function targetNs(grid, k) {
  const target =
    grid.origin + roundHalfEven(k * grid.periodNum, grid.periodDen);
  if (target < INT64_MIN || target > INT64_MAX) {
    throw new VectorVerificationError("sampling target does not fit signed int64");
  }
  return target;
}

function firstKAtOrAfter(grid, timestamp) {
  const relativeTarget = timestamp - grid.origin;
  const boundary = (2n * relativeTarget - 1n) * grid.periodDen;
  const scale = 2n * grid.periodNum;
  const [quotient, remainder] = floorDivmod(boundary, scale);
  if (remainder === 0n && relativeTarget % 2n === 0n) return quotient;
  return quotient + 1n;
}

function enumerateTargets(grid, start, end) {
  if (start >= end) {
    throw new VectorVerificationError("interval start must be less than end");
  }
  const targets = [];
  let k = firstKAtOrAfter(grid, start);
  const stopK = firstKAtOrAfter(grid, end);
  while (k < stopK) {
    const timestamp = targetNs(grid, k);
    targets.push({ k, targetNs: timestamp });
    if (targets.length > 100_000) {
      throw new VectorVerificationError("vector target budget exceeded");
    }
    k = firstKAtOrAfter(grid, timestamp + 1n);
  }
  return targets;
}

function parseFrame(value, label) {
  const fields = requireObject(value, label);
  const locatorHex = requireText(fields.source_locator_hex, `${label}.source_locator_hex`);
  if (!LOWER_HEX_OCTETS.test(locatorHex)) {
    throw new VectorVerificationError(
      `${label}.source_locator_hex must be lowercase hexadecimal octets`,
    );
  }
  if (typeof fields.decodable !== "boolean") {
    throw new VectorVerificationError(`${label}.decodable must be a boolean`);
  }
  return {
    alignedTimestampNs: int64(fields.aligned_timestamp_ns, `${label}.aligned_timestamp_ns`),
    sourceTimestampNs: int64(fields.source_timestamp_ns, `${label}.source_timestamp_ns`),
    locatorHex,
    locatorBytes: Buffer.from(locatorHex, "hex"),
    decodable: fields.decodable,
  };
}

function compareBigInt(left, right) {
  return left < right ? -1 : left > right ? 1 : 0;
}

function compareFramesForTarget(left, right, target) {
  return (
    compareBigInt(
      absolute(left.alignedTimestampNs - target.targetNs),
      absolute(right.alignedTimestampNs - target.targetNs),
    ) ||
    compareBigInt(left.alignedTimestampNs, right.alignedTimestampNs) ||
    compareBigInt(left.sourceTimestampNs, right.sourceTimestampNs) ||
    Buffer.compare(left.locatorBytes, right.locatorBytes)
  );
}

function selectFrames(targets, frames, start, end, tolerance) {
  if (tolerance < 0n) {
    throw new VectorVerificationError("selection tolerance must be nonnegative");
  }
  const intervalFrames = frames.filter(
    (frame) => start <= frame.alignedTimestampNs && frame.alignedTimestampNs < end,
  );
  const provisional = targets.map((target) => {
    const withinTolerance = intervalFrames
      .filter(
        (frame) => absolute(frame.alignedTimestampNs - target.targetNs) <= tolerance,
      )
      .sort((left, right) => compareFramesForTarget(left, right, target));
    const selected = withinTolerance.find((frame) => frame.decodable);
    if (selected !== undefined) {
      return {
        ...target,
        status: "SELECTED",
        frame: selected,
        deltaToTargetNs: selected.alignedTimestampNs - target.targetNs,
      };
    }
    if (withinTolerance.length > 0) {
      const failed = withinTolerance[0];
      return {
        ...target,
        status: "DECODE_FAILED",
        frame: failed,
        deltaToTargetNs: failed.alignedTimestampNs - target.targetNs,
      };
    }
    return {
      ...target,
      status: "NO_FRAME_WITHIN_TOLERANCE",
      frame: null,
      deltaToTargetNs: null,
    };
  });

  const assignments = new Map();
  provisional.forEach((selection, index) => {
    if (selection.status === "SELECTED") {
      const indexes = assignments.get(selection.frame.locatorHex) ?? [];
      indexes.push(index);
      assignments.set(selection.frame.locatorHex, indexes);
    }
  });
  for (const indexes of assignments.values()) {
    if (indexes.length < 2) continue;
    indexes.sort((leftIndex, rightIndex) => {
      const left = provisional[leftIndex];
      const right = provisional[rightIndex];
      return (
        compareBigInt(absolute(left.deltaToTargetNs), absolute(right.deltaToTargetNs)) ||
        compareBigInt(left.targetNs, right.targetNs) ||
        compareBigInt(left.k, right.k)
      );
    });
    for (const index of indexes.slice(1)) {
      provisional[index] = { ...provisional[index], status: "DEDUPLICATED_FRAME" };
    }
  }
  return provisional;
}

function frameProjection(frame) {
  if (frame === null) return null;
  return {
    aligned_timestamp_ns: frame.alignedTimestampNs.toString(),
    source_timestamp_ns: frame.sourceTimestampNs.toString(),
    source_locator_hex: frame.locatorHex,
    decodable: frame.decodable,
  };
}

function gridOutcome(input, label) {
  const grid = createGrid(input, label);
  const interval = requireObject(input.interval, `${label}.interval`);
  const start = int64(interval.start_ns, `${label}.interval.start_ns`);
  const end = int64(interval.end_ns, `${label}.interval.end_ns`);
  const tolerance = int64(input.selection_tolerance_ns, `${label}.selection_tolerance_ns`);
  const frames = requireArray(input.frames, `${label}.frames`).map((value, index) =>
    parseFrame(value, `${label}.frames[${index}]`),
  );
  const targets = enumerateTargets(grid, start, end);
  const selections = selectFrames(targets, frames, start, end, tolerance);
  return {
    period: {
      period_num_ns: grid.periodNum.toString(),
      period_den: grid.periodDen.toString(),
    },
    targets: targets.map((target) => ({
      k: target.k.toString(),
      target_ns: target.targetNs.toString(),
    })),
    selections: selections.map((selection) => ({
      k: selection.k.toString(),
      target_ns: selection.targetNs.toString(),
      status: selection.status,
      frame: frameProjection(selection.frame),
      delta_to_target_ns:
        selection.deltaToTargetNs === null ? null : selection.deltaToTargetNs.toString(),
    })),
  };
}

function roundingOutcome(input, label) {
  const fractions = requireArray(input.fractions, `${label}.fractions`);
  return {
    rounding_results: fractions.map((value, index) => {
      const fraction = requireObject(value, `${label}.fractions[${index}]`);
      const numerator = integer(fraction.numerator, `${label}.fractions[${index}].numerator`);
      const denominator = integer(
        fraction.denominator,
        `${label}.fractions[${index}].denominator`,
      );
      return roundHalfEven(numerator, denominator).toString();
    }),
  };
}

function requireEqual(actual, expected, label) {
  if (canonicalJson(actual) !== canonicalJson(expected)) {
    throw new VectorVerificationError(`${label} does not match`);
  }
}

function verifyVectors(path) {
  const root = requireObject(JSON.parse(readFileSync(path, "utf8")), "document");
  if (root.suite_id !== "robata-rational-grid-canonicalization") {
    throw new VectorVerificationError("unexpected suite_id");
  }
  if (root.suite_version !== "1.0.0") {
    throw new VectorVerificationError("unexpected suite_version");
  }
  if (root.canonicalization !== "RFC8785") {
    throw new VectorVerificationError("canonicalization must be RFC8785");
  }
  if (root.integer_encoding !== "canonical-base10-string") {
    throw new VectorVerificationError("integer_encoding must be canonical-base10-string");
  }

  const checked = new Set();
  for (const [index, value] of requireArray(root.vectors, "vectors").entries()) {
    const vector = requireObject(value, `vectors[${index}]`);
    const caseId = requireText(vector.case_id, `vectors[${index}].case_id`);
    if (checked.has(caseId)) {
      throw new VectorVerificationError(`duplicate case_id: ${caseId}`);
    }
    const label = `vector ${JSON.stringify(caseId)}`;
    const input = requireObject(vector.input, `${label}.input`);
    const expected = requireObject(vector.expected, `${label}.expected`);
    const canonicalInput = Buffer.from(canonicalJson(input), "utf8");
    if (expected.canonical_input_bytes_hex !== canonicalInput.toString("hex")) {
      throw new VectorVerificationError(`${label} canonical input bytes do not match`);
    }
    if (expected.canonical_input_sha256 !== sha256(canonicalInput)) {
      throw new VectorVerificationError(`${label} canonical input SHA-256 does not match`);
    }

    let outcome;
    if (vector.operation === "ROUND_HALF_EVEN") {
      outcome = roundingOutcome(input, label);
    } else if (vector.operation === "GRID_SELECTION") {
      outcome = gridOutcome(input, label);
    } else {
      throw new VectorVerificationError(
        `${label} has unsupported operation ${JSON.stringify(vector.operation)}`,
      );
    }
    requireEqual(outcome, expected.outcome, `${label} runtime outcome`);
    checked.add(caseId);
  }
  if (checked.size === 0) {
    throw new VectorVerificationError("vector suite must not be empty");
  }
  return checked.size;
}

const vectorPath = process.argv[2] === undefined ? DEFAULT_VECTOR_PATH : resolve(process.argv[2]);
const verifiedVectors = verifyVectors(vectorPath);
process.stdout.write(
  `${JSON.stringify({
    implementation: "node-bigint-independent-v1",
    suite: "robata-rational-grid-canonicalization@1.0.0",
    verified_vectors: verifiedVectors,
  })}\n`,
);

File: scripts/verify_rational_grid_vectors.py
"""Verify the language-neutral rational-grid and canonicalization golden vectors."""

from __future__ import annotations

import argparse
import json
import re
import sys
from collections.abc import Mapping
from pathlib import Path
from typing import Any

REPOSITORY_ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(REPOSITORY_ROOT / "src"))

from robata.alignment.rational_time import round_half_even  # noqa: E402
from robata.contracts.hashing import canonical_json_bytes, exact_bytes_sha256  # noqa: E402
from robata.sampling.grid import (  # noqa: E402
    FrameCandidate,
    SamplingGrid,
    SamplingRate,
    TargetSelection,
)

DEFAULT_VECTOR_PATH = REPOSITORY_ROOT / "conformance" / "rational-grid-canonicalization-v1.json"
_CANONICAL_DECIMAL = re.compile(r"^(?:0|-[1-9][0-9]*|[1-9][0-9]*)$")


class VectorVerificationError(ValueError):
    """One checked-in conformance vector is malformed or disagrees with the runtime."""


def _reject_duplicate_keys(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            raise VectorVerificationError(f"duplicate JSON object key: {key!r}")
        result[key] = value
    return result


def _object(value: Any, label: str) -> Mapping[str, Any]:
    if not isinstance(value, dict):
        raise VectorVerificationError(f"{label} must be a JSON object")
    return value


def _array(value: Any, label: str) -> list[Any]:
    if not isinstance(value, list):
        raise VectorVerificationError(f"{label} must be a JSON array")
    return value


def _text(value: Any, label: str) -> str:
    if not isinstance(value, str) or not value:
        raise VectorVerificationError(f"{label} must be a nonempty string")
    return value


def _integer(value: Any, label: str) -> int:
    text = _text(value, label)
    if _CANONICAL_DECIMAL.fullmatch(text) is None:
        raise VectorVerificationError(f"{label} must use canonical base-10 integer text")
    return int(text)


def _frame(value: Any, label: str) -> FrameCandidate:
    fields = _object(value, label)
    locator_hex = _text(fields.get("source_locator_hex"), f"{label}.source_locator_hex")
    if locator_hex.lower() != locator_hex or len(locator_hex) % 2 != 0:
        raise VectorVerificationError(f"{label}.source_locator_hex must be lowercase octets")
    try:
        locator = bytes.fromhex(locator_hex)
    except ValueError as exc:
        raise VectorVerificationError(
            f"{label}.source_locator_hex must be lowercase hexadecimal"
        ) from exc
    decodable = fields.get("decodable")
    if not isinstance(decodable, bool):
        raise VectorVerificationError(f"{label}.decodable must be a boolean")
    return FrameCandidate(
        aligned_timestamp_ns=_integer(
            fields.get("aligned_timestamp_ns"), f"{label}.aligned_timestamp_ns"
        ),
        source_timestamp_ns=_integer(
            fields.get("source_timestamp_ns"), f"{label}.source_timestamp_ns"
        ),
        source_locator_bytes=locator,
        decodable=decodable,
    )


def _target_projection(grid: SamplingGrid, start_ns: int, end_ns: int) -> list[dict[str, str]]:
    return [
        {"k": str(target.k), "target_ns": str(target.target_ns)}
        for target in grid.enumerate_targets(start_ns, end_ns)
    ]


def _selection_projection(selection: TargetSelection) -> dict[str, Any]:
    frame = selection.frame
    frame_projection = (
        None
        if frame is None
        else {
            "aligned_timestamp_ns": str(frame.aligned_timestamp_ns),
            "source_timestamp_ns": str(frame.source_timestamp_ns),
            "source_locator_hex": frame.source_locator_bytes.hex(),
            "decodable": frame.decodable,
        }
    )
    return {
        "k": str(selection.k),
        "target_ns": str(selection.target_ns),
        "status": selection.status.value,
        "frame": frame_projection,
        "delta_to_target_ns": (
            None if selection.delta_to_target_ns is None else str(selection.delta_to_target_ns)
        ),
    }


def _rounding_outcome(vector_input: Mapping[str, Any], label: str) -> dict[str, Any]:
    fractions = _array(vector_input.get("fractions"), f"{label}.fractions")
    results: list[str] = []
    for index, value in enumerate(fractions):
        fraction = _object(value, f"{label}.fractions[{index}]")
        numerator = _integer(fraction.get("numerator"), f"{label}.fractions[{index}].numerator")
        denominator = _integer(
            fraction.get("denominator"), f"{label}.fractions[{index}].denominator"
        )
        results.append(str(round_half_even(numerator, denominator)))
    return {"rounding_results": results}


def _grid_outcome(vector_input: Mapping[str, Any], label: str) -> dict[str, Any]:
    grid_fields = _object(vector_input.get("grid"), f"{label}.grid")
    interval = _object(vector_input.get("interval"), f"{label}.interval")
    grid = SamplingGrid(
        grid_origin_ns=_integer(grid_fields.get("grid_origin_ns"), f"{label}.grid.grid_origin_ns"),
        rate=SamplingRate(
            numerator=_integer(grid_fields.get("rate_num"), f"{label}.grid.rate_num"),
            denominator=_integer(grid_fields.get("rate_den"), f"{label}.grid.rate_den"),
        ),
    )
    start_ns = _integer(interval.get("start_ns"), f"{label}.interval.start_ns")
    end_ns = _integer(interval.get("end_ns"), f"{label}.interval.end_ns")
    tolerance_ns = _integer(
        vector_input.get("selection_tolerance_ns"), f"{label}.selection_tolerance_ns"
    )
    frames = tuple(
        _frame(value, f"{label}.frames[{index}]")
        for index, value in enumerate(_array(vector_input.get("frames"), f"{label}.frames"))
    )
    selections = grid.select_frames(frames, start_ns, end_ns, tolerance_ns)
    return {
        "period": {
            "period_num_ns": str(grid.period_num_ns),
            "period_den": str(grid.period_den),
        },
        "targets": _target_projection(grid, start_ns, end_ns),
        "selections": [_selection_projection(selection) for selection in selections],
    }


def verify_vectors(path: Path = DEFAULT_VECTOR_PATH) -> tuple[str, ...]:
    """Verify canonical input pins and runtime outcomes, returning checked case IDs."""

    try:
        document = json.loads(
            path.read_text(encoding="utf-8"),
            object_pairs_hook=_reject_duplicate_keys,
        )
    except (OSError, UnicodeError, json.JSONDecodeError) as exc:
        raise VectorVerificationError(f"cannot read vector document {path}: {exc}") from exc
    root = _object(document, "document")
    if root.get("suite_id") != "robata-rational-grid-canonicalization":
        raise VectorVerificationError("unexpected suite_id")
    if root.get("suite_version") != "1.0.0":
        raise VectorVerificationError("unexpected suite_version")
    if root.get("canonicalization") != "RFC8785":
        raise VectorVerificationError("canonicalization must be RFC8785")
    if root.get("integer_encoding") != "canonical-base10-string":
        raise VectorVerificationError("integer_encoding must be canonical-base10-string")

    checked: list[str] = []
    for index, value in enumerate(_array(root.get("vectors"), "vectors")):
        vector = _object(value, f"vectors[{index}]")
        case_id = _text(vector.get("case_id"), f"vectors[{index}].case_id")
        if case_id in checked:
            raise VectorVerificationError(f"duplicate case_id: {case_id}")
        label = f"vector {case_id!r}"
        vector_input = _object(vector.get("input"), f"{label}.input")
        expected = _object(vector.get("expected"), f"{label}.expected")

        canonical_input = canonical_json_bytes(vector_input)
        if expected.get("canonical_input_bytes_hex") != canonical_input.hex():
            raise VectorVerificationError(f"{label} canonical input bytes do not match")
        if expected.get("canonical_input_sha256") != exact_bytes_sha256(canonical_input):
            raise VectorVerificationError(f"{label} canonical input SHA-256 does not match")

        operation = vector.get("operation")
        if operation == "ROUND_HALF_EVEN":
            actual_outcome = _rounding_outcome(vector_input, label)
        elif operation == "GRID_SELECTION":
            actual_outcome = _grid_outcome(vector_input, label)
        else:
            raise VectorVerificationError(f"{label} has unsupported operation {operation!r}")
        if expected.get("outcome") != actual_outcome:
            raise VectorVerificationError(f"{label} runtime outcome does not match")
        checked.append(case_id)
    if not checked:
        raise VectorVerificationError("vector suite must not be empty")
    return tuple(checked)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("path", nargs="?", type=Path, default=DEFAULT_VECTOR_PATH)
    args = parser.parse_args(argv)
    checked = verify_vectors(args.path)
    print(
        json.dumps(
            {
                "implementation": "python-authoritative-runtime-v1",
                "suite": "robata-rational-grid-canonicalization@1.0.0",
                "verified_vectors": len(checked),
            },
            separators=(",", ":"),
            sort_keys=True,
        )
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
